aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -753,7 +753,7 @@ Tensor &all_out(Tensor &result, const Tensor &self, int64_t dim, bool keepdim) {
     return result;
   } else {
     auto iter = make_reduction(
-      "all", result, self, dim, keepdim, self.scalar_type());
+      "all", result, self, dim, keepdim, /*out_dtype=*/kBool);
```

Comment:
this invocation of make_reduction will explicitly cast input to bool type, I doubt that's what you want.

Reply:
I think we need that for CPU dynamic casting is not supported,
https://github.com/pytorch/pytorch/blob/0652d755d3a5d6c21a695ff9b57d8be972496a4f/aten/src/ATen/native/cpu/Reduce.h#L247-L254

And the inferred type of the data, is based on the return type of `op` passed to `binary_kernel_reduce_vec`,
https://github.com/pytorch/pytorch/blob/0652d755d3a5d6c21a695ff9b57d8be972496a4f/aten/src/ATen/native/cpu/Reduce.h#L35-L36

https://github.com/pytorch/pytorch/blob/0652d755d3a5d6c21a695ff9b57d8be972496a4f/aten/src/ATen/native/cpu/Reduce.h#L13-L17

Let me know if I am missing something :thinking: 
Thanks!

Reply:
It should be possible to not always cast the inputs to bool. I think there are two cases:

- both inputs have the same scalar type
- the inputs have different scalar types

For the first case the conversion to bool can happen in the kernel.

For the second case the inputs do need to be copied to avoid excessive template instantiations. In this case the inputs might as well be cast to bool ahead of the kernel being called.

The first case, where the scalar types are the same, is probably much more common than different scalar types, and would be nice to optimize for. Would it be possible to address that case?

Reply:
> The first case, where the scalar types are the same, is probably much more common than different scalar types, and would be nice to optimize for. Would it be possible to address that case?

This is already taken care of,
https://github.com/pytorch/pytorch/blob/72918e475e40a26cbf56f4fc2641e7c3e7365f8a/aten/src/ATen/native/ReduceOpsUtils.h#L173-L194

At the bottom there is
https://github.com/pytorch/pytorch/blob/72918e475e40a26cbf56f4fc2641e7c3e7365f8a/aten/src/ATen/native/ReduceOpsUtils.h#L190-L192